### ISSUE TITLE
Possible spreadsheet migration tool

### ISSUE THREAD

**Issue by ESapenaVentura** [HumanCellAtlas/ingest-central#431]
# tl;dr
A spreadsheet migration tool is needed. No need to develop it all at once; it could begin as a spreadsheet check. There are several options to develop the tool, but in the end all of them should rely on `property_migrations`. This tool could be ultimately integrated into the ingest ui.  **Help is needed with this in order to not duplicate efforts**.

# Why we need this tool
**EDIT 07/06/2019:** @daniwelter is developing a tool that covers most of the requirements. Ticket can be found [here](https://github.com/HumanCellAtlas/schema-template-generator/issues/5) [WIP].

Following what was talked in the wrangler catch-up meeting (06/06/2019), and also following @cjvillar concerns on the time loss due to manual metadata spreadsheet migrations (Specially in the use case of depending on a contributor to fill up the spreadsheet, which can take months), it was proposed to develop an automatic tool for migration. 

# Planning of tool development
This tool can have several facets to it: 
- It could began as a simple checker (Creating a log file) that tells you what's wrong with the metadata spreadsheet, and so the wrangler knows exactly where to look at when updating the spreadsheet. Right now, the only way to do it is through uploading the spreadsheet to ingest, which is not optimal.
- When fully developed, its function in ingest could be to detect if the spreadsheet is up-to-date (Only major changes, the ones without backwards compatibility), detect the changes, make those changes to the spreadsheet and, if necessary, give the updated spreadsheet back to the contributor. 

In order to be able to make it 

# Possible workflow of the tool
1. The contributor presents the tool with a spreadsheet (Unknown if it's outdated or not).
1. At this point, we actually have 3 options:
   1. Convert the spreadsheet back to json to compare against current json schema. This would require a tool to do such conversion, and somehow store the data that is in the spreadsheet.
   1. Check the spreadsheet against the json schema. This would require a tool that identifies where the column in the spreadsheet is coming from, and looks on that schema for the comparison.
   1. Convert the current json_schema (full) to a spreadsheet and, for each tab, compare 4th row of each of the columns from the outdated spreadsheet against the full, up-to-date schema. 
1. Save the changes.

This tool would rely on the `property_migrations.json` file to decide what to do with the fields that don't match. If the change is not recorded there (e.g., the spreadsheet was created before property_migrations, or plain human error of not recording a breaking change to the file), it could record it in a log and throw an error at the end.

# Reason to chose this repo for the tool
@lauraclarke, @zperova, @malloryfreeberg  and I have discussed over which one is the best repo to have this tool on, and we think that `ingest-central` is the most suited for it. This kind of tools are not suited for the `metadata-schema` repo, and even though we thought that the `schema-template-generator` repo would actually be the most suited for it, right now is not very well known. 

# Further help needed
- We need to make sure that we are not duplicating efforts. Everyone that has worked on any kind of script that would help with this tool should communicate it here.
- This is just a draft, so any comment/suggestion is welcome. Pings to people that could help are, too!
- @daniwelter, @justincc, @simonjupp, @HumanCellAtlas/ingest, @HumanCellAtlas/wranglers,  ideas on this?

**EDIT 07/06/2019: Added new updates from @daniwelter about this tool**

**Comment by daniwelter** [comment]
What's wrong with the tool I developed 3 weeks ago?

**Comment by ESapenaVentura** [comment]
I'm really sorry, I didn't know you were developing a tool to automatically update the spreadsheets. That's wonderful to hear! Could you point me to the ticket? :) 

(PS: This is partially why this ticket was needed, so we can acknowledge if someone in the team is actually working in this!)

**Comment by daniwelter** [comment]
No one filed a ticket, although I think Norman recorded it in the sprint priorities. Branch here https://github.com/HumanCellAtlas/schema-template-generator/tree/spreadsheet_migration

**Comment by ESapenaVentura** [comment]
Thanks for the link to the branch! I'll take a look at it. 

Well, I'm sure either @lauraclarke or @malloryfreeberg could explain this better than me, but we had a talk in which we thought about this tool being used by ingest to provide the future data contributor with an updated spreadsheet automatically. 

I'm going to update the ticket with this new info :)

**Comment by malloryfreeberg** [comment]
@daniwelter if there isn't a ticket, could you please make one? We need to be able to track this work,  know what the use cases are that are being addressed, and what features are being implemented. Thanks.

**Comment by daniwelter** [comment]
@malloryfreeberg https://github.com/HumanCellAtlas/schema-template-generator/issues/5

Ticket can't go into ingest-central as I'm working on it and I'm not supposed to have ticket in ingest-central as I'm not a dev.
